core/state_processor_test.go

```diff
@@ -616,3 +620,838 @@ func getParentBlockHash(statedb *state.StateDB, number uint64) common.Hash {
+		// must be nil.
+		if balanceStateDiff.NewValue != nil {
+			t.Fatalf("the post-state balance after self-destruct must be empty")
+		}
+	}
+}
+
+func TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary(t *testing.T) {
+	var (
+		config     = *testKaustinenLikeChainConfig
+		signer     = types.LatestSigner(&config)
+		testKey, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		coinbase   = common.HexToAddress("0x71562b71999873DB5b286dF957af199Ec94617F7")
+		account1   = common.HexToAddress("0x687704DB07e902e9A8B3754031D168D46E3D586e")
+		account2   = common.HexToAddress("0x6177843db3138ae69679A54b95cf345ED759450d")
+		gspec      = &Genesis{
+			Config: &config,
+			Alloc: GenesisAlloc{
+				coinbase: GenesisAccount{
+					Balance: big.NewInt(1000000000000000000), // 1 ether
+					Nonce:   0,
+				},
+				account1: GenesisAccount{
+					Balance: big.NewInt(1000000000000000000), // 1 ether
+					Nonce:   0,
+				},
+				account2: GenesisAccount{
+					Balance: big.NewInt(1000000000000000000), // 1 ether
+					Nonce:   3,
+				},
+				params.HistoryStorageAddress: GenesisAccount{
+					Balance: big.NewInt(0),
+					Nonce:   1,
+				},
+			},
+		}
+	)
+
+	// The test txs were taken from a secondary testnet with chain id 69421
+	config.ChainID.SetUint64(69421)
+
+	// The goal of this test is to test SELFDESTRUCT that happens in a contract execution which is created
+	// in **the same** transaction sending the remaining balance to itself.
+
+	selfDestructContract := []byte{
+		byte(vm.PUSH20),
+		0x3a, 0x22, 0x0f, 0x35, 0x12, 0x52, 0x08, 0x9d, 0x38, 0x5b, 0x29, 0xbe, 0xca, 0x14, 0xe2, 0x7f, 0x20, 0x4c, 0x29, 0x6a, // 0x3a220f351252089d385b29beca14e27f204c296a
+		byte(vm.SELFDESTRUCT),
+	}
+	selfDestructContractAddr := common.HexToAddress("3a220f351252089d385b29beca14e27f204c296a")
+	_, _, _, _, statediffs := GenerateVerkleChainWithGenesis(gspec, beacon.New(ethash.NewFaker()), 1, func(i int, gen *BlockGen) {
+		gen.SetPoS()
+		tx, _ := types.SignTx(types.NewContractCreation(0, big.NewInt(42), 100_000, big.NewInt(875000000), selfDestructContract), signer, testKey)
+		gen.AddTx(tx)
+	})
+
+	{ // Check self-destructed contract in the witness
+		selfDestructContractTreeKey := utils.CodeHashKey(selfDestructContractAddr[:])
+
+		var stateDiffIdx = -1
+		for i, stemStateDiff := range statediffs[0] {
+			if bytes.Equal(stemStateDiff.Stem[:], selfDestructContractTreeKey[:31]) {
+				stateDiffIdx = i
+				break
+			}
+		}
+		if stateDiffIdx == -1 {
+			t.Fatalf("no state diff found for stem")
+		}
+
+		balanceStateDiff := statediffs[0][stateDiffIdx].SuffixDiffs[0]
+		if balanceStateDiff.Suffix != utils.BasicDataLeafKey {
+			t.Fatalf("balance invalid suffix")
+		}
+
+		if balanceStateDiff.CurrentValue != nil {
+			t.Fatalf("the pre-state balance before must be nil, since the contract didn't exist")
+		}
+
+		if balanceStateDiff.NewValue != nil {
```

Comment:
If I understand correctly, the situation here is:

- Create contract along with 42 wei sent
- self-destruct the new contract by itself
- Beneficiary is the new contract itself
- the 42 wei is burnt

If so, please add some test description as the function `TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary ` comment. It will be very beneficial for reviewing